pennylane_lightning/src/StateVector.hpp

```diff
@@ -15,19 +15,717 @@
  * Defines the class representation for quantum state vectors.
  */
+
 #pragma once
 
-#include "typedefs.hpp"
+#include <cmath>
+#include <complex>
+#include <functional>
+#include <set>
+#include <stdexcept>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "Util.hpp"
+
+namespace {
+using namespace std::placeholders;
+using std::bind;
+using std::size_t;
+using std::string;
+using std::vector;
+}; // namespace
 
 namespace Pennylane {
 
-class StateVector {
+template <class fp_t = double> class StateVector {
+  private:
+    using CFP_t = std::complex<fp_t>;
+
+    using Func =
+        std::function<void(const vector<size_t> &, const vector<size_t> &, bool,
+                           const vector<fp_t> &)>;
+
+    using FMap = std::unordered_map<string, Func>;
+
+    static constexpr CFP_t ONE = {1, 0};
+    static constexpr CFP_t ZERO = {0, 0};
+    static constexpr CFP_t IMAG = {0, 1};
+    inline static const CFP_t SQRT2 = {static_cast<fp_t>(std::sqrt(2)), 0};
+    inline static const CFP_t INVSQRT2 = {static_cast<fp_t>(1 / std::sqrt(2)),
+                                          0};
+
+    const FMap gates_;
+    const std::unordered_map<string, size_t> gate_wires_;
+
+    CFP_t *const arr_;
+    const size_t length_;
+    const size_t num_qubits_;
+
+    vector<size_t>
+    getIndicesAfterExclusion(const vector<size_t> &indicesToExclude) {
+        std::set<size_t> indices;
+        for (size_t i = 0; i < num_qubits_; i++) {
+            indices.emplace(i);
+        }
+        for (const size_t &excludedIndex : indicesToExclude) {
+            indices.erase(excludedIndex);
+        }
+        return {indices.begin(), indices.end()};
+    }
+
+    vector<size_t> generateBitPatterns(const vector<size_t> &qubitIndices) {
+        vector<size_t> indices;
+        indices.reserve(Util::exp2(qubitIndices.size()));
+        indices.emplace_back(0);
+        for (int i = qubitIndices.size() - 1; i >= 0; i--) {
+            size_t value =
+                Util::maxDecimalForQubit(qubitIndices[i], num_qubits_);
+            size_t currentSize = indices.size();
+            for (size_t j = 0; j < currentSize; j++) {
+                indices.emplace_back(indices[j] + value);
+            }
+        }
+        return indices;
+    }
 
   public:
-    CplxType *const arr;
-    const size_t length;
+    StateVector()
+        : arr_{nullptr}, length_{0}, num_qubits_{0}, gate_wires_{}, gates_{} {};
+    StateVector(CFP_t *arr, size_t length)
+        : arr_{arr}, length_{length}, num_qubits_{Util::log2(length_)},
+          gate_wires_{
+              {"PauliX", 1}, {"PauliY", 1}, {"PauliZ", 1},     {"Hadamard", 1},
+              {"T", 1},      {"S", 1},      {"RX", 1},         {"RY", 1},
+              {"RZ", 1},     {"Rot", 1},    {"PhaseShift", 1}, {"CNOT", 2},
+              {"SWAP", 2},   {"CZ", 2},     {"CRX", 2},        {"CRY", 2},
+              {"CRZ", 2},    {"CRot", 2},   {"CSWAP", 3},      {"Toffoli", 3}},
```

Comment:
Sure thing. The previous implementation had some logic to check the gate classes for the number of supported wires to act upon. Since the classes are now methods, the mapping of a gate label to supported wires needed to be elsewhere. I created a private dictionary to map between gate labels and the number of supported wires (`gate_wires_`). This allows minimal changes to the apply methods to match the older functionality. 

Since the dictionary is declared as `const` it must be initialised with this initialiser list syntax. Though, we can move these to a `inline static const / constexpr` and put it outside the class instantiation, though for now I think this is fine.

Reply:
Thanks!